torch/utils/data/dataloader.py

```diff
@@ -485,14 +523,72 @@ def __iter__(self) -> _BaseDataLoaderIter:
         # However, in the case of a multiple workers iterator
         # the iterator is only created once in the lifetime of the
         # DataLoader object so that workers can be reused
-        if self.persistent_workers and self.num_workers > 0:
+        if self._initial_iter_for_state_dict:
+            self._initial_iter_for_state_dict = False
+            assert self._iterator is not None
+        elif self.persistent_workers and self.num_workers > 0:
             if self._iterator is None:
                 self._iterator = self._get_iterator()
             else:
                 self._iterator._reset(self)
-            return self._iterator
         else:
-            return self._get_iterator()
+            self._iterator = self._get_iterator()
+
+        if (
+            self.stateful
+            and self._iterator is not None
+            and getattr(self._iterator, "_finished", False)
+        ):
+            self._iterator = self._get_iterator()
+
+        return self._iterator
+
+    def state_dict(self) -> dict[str, Any]:
+        """
+        Return the state of the dataloader for checkpointing purposes.
+
+        Raises:
+            RuntimeError: If the dataloader was not created with stateful=True
+        """
+        if not self.stateful:
+            raise RuntimeError(
+                "DataLoader.state_dict() requires stateful=True. "
+                "Create DataLoader with DataLoader(..., stateful=True)"
+            )
+
+        if self._iterator is None:
+            self._iterator = self._get_iterator()
+            self._initial_iter_for_state_dict = True
+
+        if hasattr(self._iterator, "state_dict"):
+            return self._iterator.state_dict()
+        else:
+            raise RuntimeError(
+                "DataLoader iterator does not support state_dict(). "
+                "This should not happen with stateful=True."
+            )
+
+    def load_state_dict(self, state_dict: dict[str, Any]) -> None:
+        """
+        Load the state of the dataloader from a checkpoint.
+
+        Args:
+            state_dict: The state dictionary returned by state_dict()
+
+        Raises:
+            RuntimeError: If the dataloader was not created with stateful=True
+        """
+        if not self.stateful:
+            raise RuntimeError(
+                "DataLoader.load_state_dict() requires stateful=True. "
+                "Create DataLoader with DataLoader(..., stateful=True)"
+            )
+
+        self._iterator = None
+        self._initial_iter_for_state_dict = False
+        if state_dict == {}:
+            return
+        self.next_iter_state = state_dict
```

Comment:
Something is weird here.  

Not sure what the expected behaviour is when passing an empty dict.  IMO it should be one of the following (and we should document which one):
1. This is not allowed, throw an exception
2. This is a no-op.  If this is the case, then all of the statements here should go inside the `if` block
3. This means "reset the state to the very beginning".  If this is the case then we should be setting `self.next_iter_state = None`.